gateway/scim.py

```diff
@@ -47,62 +47,75 @@ def check_apikey(*args, **kwargs):
 
 @scim_app.route('/Devices', methods=['POST'])
 @authenticate_user
-def scim_addusers():
+def add_scim_entry():
     """
     This code defines a SCIM API endpoint for creating users,
     extracts data from the request JSON, and stores user information in a database.
     """
     # Get the request json and check if it is valid
     if not request.json:
-        return make_response(
-            jsonify(
-                {
-                    "schemas": ["urn:ietf:params:scim:schemas:core:2.0:Error"],
-                    "scimType": "invalidSyntax",
-                    "detail": "Request body is not valid JSON.",
-                    "status": 400,
-                }
-            ),
-            400,
-        )
+        return blow_an_error("Request body is not valid JSON.",400)
 
     schemas = request.json.get("schemas")
+    device_id=request.json.get("id")
+
+    if device_id:
+        return blow_an_error("Specifying id on create not permitted", 400)
+
+    device_id = uuid.uuid4()
+
+    endpoint_apps = None
+    appextschema = \
+        'urn:ietf:params:scim:schemas:extension:endpointAppsExt:2.0:Device'
+    endpoint_apps_ext = request.json.get(appextschema, None)
+    if endpoint_apps_ext:
+        if not appextschema in schemas:
+            schemas.append(appextschema)
+        applications = endpoint_apps_ext.get("applications")
+        endpoint_app_ids = [app.get("value") for app in applications]
+        # Select all endpoint apps from the database
+        endpoint_apps = session.scalars(select(EndpointApp).filter(
+            EndpointApp.id.in_(endpoint_app_ids))).all()
+
+
+    # Add device core object
+    try:
+        entry = Device(
+            schemas=schemas,
+            device_display_name=request.json["deviceDisplayName"],
+            admin_state=request.json["adminState"],
+            endpoint_apps=endpoint_apps,
+            created_time=datetime.datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
+        )
+        if 'urn:ietf:params:scim:schemas:extension:ble:2.0:Device' in schemas:
+            entry.ble_extension = ble_create_device(request,device_id)
+        # Dispatch to appropriate function
+        session.add(entry)
+        session.commit()
 
-    # Dispatch to appropriate function
-    if 'urn:ietf:params:scim:schemas:extension:ble:2.0:Device' in schemas:
-        return ble_create_device(request)
-    return make_response(
-        jsonify(
-            {
-                "schemas": ["urn:ietf:params:scim:schemas:core:2.0:Error"],
-                "detail": "Extension not implemented.",
-                "status": 501
-            }
-        ),
-        501
-    )
+        core=entry.serialize()
+
+    except DeviceExists:
+        return blow_an_error("Device already exists", 500)
```

Comment:
This should be a HTTP status code 409 (Conflict) with a "scimType" error code of "uniqueness" according to RFC 7644 (section 3.3).

Reply:
Ok, we'll need to review all of the scimTypes...